ably/realtime_channel_spec_test.go

```diff
@@ -155,6 +157,1258 @@ func attachAndWait(t *testing.T, channel *ably.RealtimeChannel) {
+		if expected, got := ably.ChannelStateAttaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateFailed, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		attachErr := <-channelTransitioner.err[chAttaching]
+
+		if attachErr == nil {
+			t.Fatal("attach should return channel failed error")
+		}
+
+		if ably.UnwrapErrorCode(attachErr) != 50001 {
+			t.Fatalf("want code=50001; got %d", ably.UnwrapErrorCode(err))
+		}
+
+		ablytest.Instantly.NoRecv(t, nil, channelStateChanges, t.Fatalf) // Should not make any change to the channel state
+	})
+
+	t.Run("RTL4d : should return error on DETACHED while attaching channel", func(t *testing.T) {
+		t.Parallel()
+
+		app, err := ablytest.NewSandbox(nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer safeclose(t, app)
+
+		recorder := ablytest.NewMessageRecorder()
+		c, close := TransitionConn(t, recorder.Dial, app.Options()...)
+
+		defer safeclose(t, close)
+
+		close = c.To(
+			connecting,
+			connected,
+		)
+
+		defer safeclose(t, close)
+
+		channelTransitioner := c.Channel("test")
+		channel := channelTransitioner.Channel
+
+		channelStateChanges := make(ably.ChannelStateChanges, 10)
+		channel.OnAll(channelStateChanges.Receive)
+
+		channelTransitioner.To(chAttaching, chDetached)
+
+		// Check that the attach message is sent
+		checkIfAttachSent := recorder.CheckIfSent(proto.ActionAttach, 1)
+		if attachSent := ablytest.Instantly.IsTrue(checkIfAttachSent); !attachSent {
+			t.Fatalf("Should sent attach message, since connection state is connected")
+		}
+
+		var channelStatechange ably.ChannelStateChange
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateDetached, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		attachErr := <-channelTransitioner.err[chAttaching]
+
+		if attachErr == nil {
+			t.Fatal("attach should return channel detached error")
+		}
+
+		if ably.UnwrapErrorCode(attachErr) != 50001 {
+			t.Fatalf("want code=50001; got %d", ably.UnwrapErrorCode(err))
+		}
+
+		ablytest.Instantly.NoRecv(t, nil, channelStateChanges, t.Fatalf) // Should not make any change to the channel state
+	})
+
+	// Transition to channelStateSuspended not implemented, so cannot implement test as of now
+	t.Skip("RTL4d : should return error on SUSPENDED while attaching channel")
```

Comment:
This should be removed now.

Reply:
I have marked it as skipped, in the near future we surely have to implement it, so implementation is already there for the test, as soon as we have suspended state for channel, the test will work out of the box